14.09.2023 issue-52: insert chat id, token, token alias into database if token is valid

iColdLight opened the pull request:
Added dependency for json web token in the pom.xml file
Added validateToken method in the TokenController class
Modified registerToken method to include validation of token in the TokenController class

<!-- start pr-codex -->

---

## PR-Codex overview
### Focus of the PR:
This PR focuses on adding token validation and insertion functionality to the codebase.

### Detailed summary:
- Added `TokenValidation` class in `io.blamer.hub.pg` package.
- `TokenValidation` implements the `Tokens` interface.
- Added `validateAndInsertToken` method in `TokenValidation` to validate and insert tokens.
- Added `validateToken` method in `TokenValidation` to validate tokens.
- Added `checkIfTokenExists` method in `TokenValidation` to check if a token already exists in the database.
- Added SQL query to check if a token exists in the `checkIfTokenExists` method.
- Added `TokenAlreadyExists` exception in `io.blamer.hub.exceptions` package.
- Added `TokenInsertion` class in `io.blamer.hub.pg` package to handle token insertion.
- `TokenInsertion` implements the `Tokens` interface.
- Added `add` method in `TokenInsertion` to add tokens.
- Modified `TokenController` class in `io.blamer.hub.controller` package.
- Modified the `TokenController` constructor to accept a `Tokens` object with `@Qualifier("tokenValidation")`.
- Modified the `registerToken` method in `TokenController` to use the `tokens` object to add tokens.

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

<!-- end pr-codex -->

iColdLight committed:
issue-52 14.09.2023

zoeself commented:
@iColdLight thank you for your Pull Request. I'll assign someone to review it soon. 

 If this PR solves a ``todo`` from the code, please don't forget to remove it.

zoeself commented:
@h1alexbel please review this Pull Request. Deadline (when it should be merged or closed) is ``2023-09-17T19:18:59.313614``.

You should check if the requirements have been implemented (partially or in full), if there are unit tests covering the changes and if the CI build passes. Feel free to reject the PR or ask for changes if it's too big or not clear enough.

Estimation here is ``30 minutes``, that's how much you will be paid. You will be paid even if this PR gets rejected.

h1alexbel commented on pom.xml:
```diff
@@ -159,6 +160,11 @@ SOFTWARE.
       <artifactId>reactor-test</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
```
@iColdLight sort your dependencies, compile/provided scope first,
and only then test one

h1alexbel commented on src/main/java/io/blamer/hub/controller/TokenController.java:
```diff
@@ -41,8 +45,23 @@ public TokenController(final Tokens tkns) {
-  Mono<Void> registerToken(@RequestBody final RequestToken request)
-    throws Exception {
-    return this.tokens.add(new TokenToAdd(request).value());
+  Mono<Void> registerToken(@RequestBody final RequestToken request) throws Exception {
```
@iColdLight I think, method should be `public`

h1alexbel commented on src/main/java/io/blamer/hub/controller/TokenController.java:
```diff
@@ -41,8 +45,23 @@ public TokenController(final Tokens tkns) {
-    throws Exception {
-    return this.tokens.add(new TokenToAdd(request).value());
+  Mono<Void> registerToken(@RequestBody final RequestToken request) throws Exception {
+    String token = request.getValue();
```
@iColdLight make all your variables `final`

h1alexbel commented on src/main/java/io/blamer/hub/controller/TokenController.java:
```diff
@@ -41,8 +45,23 @@ public TokenController(final Tokens tkns) {
+  Mono<Void> registerToken(@RequestBody final RequestToken request) throws Exception {
+    String token = request.getValue();
+    String secretKey = "your-secret-key";
+
```
@iColdLight empty lines mean code smell, read [this](https://www.yegor256.com/2014/11/03/empty-line-code-smell.html)

h1alexbel reviewed (changes requested):
@iColdLight the ticket was to insert chat id, token, token alias and other stuff (can be just a placeholder for now) if the provided token is valid.
And this token is **not** JWT, it's a just simple GitHub token, simple string

WDYT?

zoeself commented:
@h1alexbel Don't forget to close this ticket before the deadline (2023-09-17T19:18:59). You are past the first half of the allowed period.

iColdLight commented:
Does it mean that I need to work on creating JWT TokenProvider, JWT TokenFIlter, JWT User, etc.?

h1alexbel commented:
@iColdLight we don't need JWTs. We are dealing with simple text token from GitHub
> And this token is not JWT, it's a just simple GitHub token, simple string

zoeself commented:
@h1alexbel Looks like you've missed the task deadline (2023-09-17T19:18:59). You are now resigned from this task.

Please stop working on it, you will not be paid. I will assign it to someone else soon.

zoeself commented:
@h1alexbel I couldn't find any assignee for this task. This is either because there are no contributors with role ``REV`` available or because the project does not have enough funds.

Please, make sure there is at least one available contributor with the required role and the project can afford to pay them.

iColdLight committed:
issue-52 17.09.2023

iColdLight committed:
issue-52 17.09.2023-2

iColdLight committed:
issue-52 17.09.2023-3

iColdLight commented:
Hi!
I made new changes. I wonder, if you can see them.
And sorry for 3 commits in a row... 
made slight changes in the last two.

l3r8yJ commented:
@iColdLight it's okay with commits, but you can read [this](https://cbea.ms/git-commit/)

l3r8yJ commented:
@h1alexbel reminder

h1alexbel commented on src/main/java/io/blamer/hub/controller/TokenController.java:
```diff
@@ -19,30 +20,30 @@
-   * Tokens.
-   */
-  private final Tokens tokens;
+    /**
```
@iColdLight don't change the layout, please

h1alexbel commented on src/main/java/io/blamer/hub/exceptions/TokenAlreadyExists.java:
```diff
@@ -0,0 +1,7 @@
+package io.blamer.hub.exceptions;
+
+public class TokenAlreadyExists extends RuntimeException{
```
@iColdLight class can be final

h1alexbel commented on src/main/java/io/blamer/hub/pg/TokenValidated.java:
```diff
@@ -0,0 +1,57 @@
+import org.springframework.stereotype.Component;
+import reactor.core.publisher.Mono;
+@Component
+public class TokenValidated implements Tokens {
```
@iColdLight It's better to name this class just like `Validated`, since your interface already saying that this is Tokens

iColdLight replied:
I didn't name my class Validated because we already have a class with that name

h1alexbel commented on src/main/java/io/blamer/hub/pg/TokenValidated.java:
```diff
@@ -0,0 +1,57 @@
+
+    @Override
+    public Mono<Void> add(final Token token) {
+        return validateToken(token)
```
@iColdLight 
This logic can be decorated, presented as at least 2 classes: first where you just insert token, and the second one where you check token validity and integrity, and then insert into database (reuse first object)
What do you think?

h1alexbel reviewed (changes requested):
@iColdLight check my comments, please

iColdLight committed:
issue-52 24.09.2023
Added TokenInsertion class
Reworked TokenValidation class
Made TokenAlreadyExists final

l3r8yJ commented:
/codex Could there be a problem here in terms of non-blocking interaction?

l3r8yJ commented:
@iColdLight take note of this comment

> > Could there be a problem here in terms of non-blocking interaction?
> 
> @l3r8yJ Yes, there could be a problem with non-blocking interaction in the code. The `TokenValidation` class implements the `Tokens` interface and overrides the `add` method. Inside the `add` method, there are multiple database interactions (`validateToken`, `checkIfTokenExists`, and `this.tokenInsertion.add`) that return `Mono` objects, indicating asynchronous operations.
> 
> However, the code does not utilize reactive operators like `flatMap`, `map`, or `zip` to chain these operations together and ensure non-blocking execution. Instead, it directly calls these methods synchronously, which may lead to blocking behavior and potential performance issues.
> 
> To ensure non-blocking interaction, the code should use reactive operators to compose the asynchronous operations and handle the results asynchronously.

it's an AI response, could be wrong

iColdLight committed:
issue-52 27.09.2023
Reworked validateAndInsertToken method in TokenValidation class

iColdLight commented:
I made a change to my TokenValidation class to ensure non-blocking interaction. May you take a look?

h1alexbel reviewed (approved):
@iColdLight core logic is fine, now is needed to pass the merge pipeline

h1alexbel commented:
@l3r8yJ take a look, please
